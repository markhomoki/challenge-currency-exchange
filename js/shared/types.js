export const EUR = 'EUR';
export const GBP = 'GBP';
export const USD = 'USD';

export const CURRENCIES = {
	[EUR]: '€',
	[GBP]: '£',
	[USD]: '$',
};
