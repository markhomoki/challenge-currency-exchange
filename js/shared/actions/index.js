import api from 'app/utils/api';

export const exchangeMoney = (fromCur, toCur, fromVal, toVal) => {
	return {
		type: 'WALLET_EXCHANGE_MONEY',
		payload: {
			fromCur,
			toCur,
			fromVal,
			toVal,
		},
	};
};

// Custom Thunk Middleware
export const rejectData = (payload, res) => {
	return {
		type: `${payload.type}_REJECTED`,
		payload: res,
	};
};

export const requestData = (payload) => {
	return {
		type: `${payload.type}_PENDING`,
	};
};

export const receiveData = (payload, res) => {
	return {
		payload: res,
		type: `${payload.type}_FULFILLED`,
	};
};

export const fetchData = payload => (dispatch) => {
	dispatch(requestData(payload));
	return api.get(payload.url)
		.then((res) => {
			if (res.status !== 200) {
				console.log('err', res);
				return;
			}
			dispatch(receiveData(payload, res));
		});
};

export const fetchRates = (fromCurrency, toCurrency) => {
	return fetchData({
		type: 'RATE',
		url: `/latest?base=${fromCurrency}&symbols=${toCurrency}`,
	});
};
