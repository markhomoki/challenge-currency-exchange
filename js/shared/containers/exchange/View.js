import * as React from 'react';
import { connect } from 'react-redux';
import { fetchRates } from 'app/actions';
import { ExchangeBox, Separator } from 'app/components';
import { GBP, USD } from 'app/types';

@connect((state) => {
	return {
		rate: state.rate,
		wallet: state.wallet,
	};
})
export default class View extends React.PureComponent {

	constructor(props) {
		super(props);

		this.state = {
			from: {
				currency: USD,
				value: 0,
				balance: props.wallet[USD],
			},
			to: {
				currency: GBP,
				value: 0,
				balance: props.wallet[GBP],
			},
		};
	}

	componentDidMount() {
		this.fetchData();
		setInterval(this.fetchData, 10000);
	}

	fetchData = () => {
		const { from, to } = this.state;

		this.props.dispatch(fetchRates(from.currency, to.currency));
	}

	exchange = (value, direction) => {

		const { rate } = this.props;

		if (!value) {
			return 0;
		}

		const formattedValue = parseFloat(value);
		const pairValue = direction === 'from' ? formattedValue * rate.value : formattedValue / rate.value;

		return parseFloat(pairValue.toFixed(2));
	}

	swap = () => {
		this.setState((prevState) => {
			return {
				from: prevState.to,
				to: prevState.from,
			};
		}, this.fetchData);
	}

	handleInputChange = (value, direction) => {
		const isFrom = direction === 'from';
		const pairValue = this.exchange(value, direction);

		const fromValue = isFrom ? value : pairValue;
		const toValue = !isFrom ? value : pairValue;

		this.setState((prevState) => {
			return {
				from: {
					...prevState.from,
					value: fromValue,
				},
				to: {
					...prevState.to,
					value: toValue,
				},
			};
		});
	}

	handleCurrencyChange = (currency, direction) => {
		const otherDir = direction === 'from' ? 'to' : 'from';

		// if matches with the pair currency then swap them
		if (this.state[otherDir].currency === currency) {
			this.swap();
			return;
		}

		this.setState((prevState) => {
			return {
				[direction]: {
					...prevState[direction],
					currency,
					balance: this.props.wallet[currency],
				},
			};
		}, this.fetchData);
	}

	render() {

		const {
			rate,
			wallet,
		} = this.props;

		const {
			from,
			to,
		} = this.state;

		const isInsufficient = wallet[from.currency] < from.value;
		const disabledClass = (!from.value || isInsufficient) ? 'disabled' : '';
		const insufficientClass = isInsufficient ? 'insufficient' : '';

		return (
			<div className="site-inner">
				<div className="exchange">
					<div className="exchange-header">
						<i className="ion-close" />
						<h1 className="page-title">Exchange</h1>
						<i className="ion-help-circled" />
					</div>
					<ExchangeBox
						{...from}
						testID="ExchangeBoxFrom"
						className={`first ${insufficientClass}`}
						direction="from"
						onInputChange={this.handleInputChange}
						onCurrencyChange={this.handleCurrencyChange}
					/>
					<Separator
						fromCurrency={from.currency}
						onSwap={this.swap}
						rate={rate.value}
						toCurrency={to.currency}
					/>
					<ExchangeBox
						{...to}
						className="alt"
						direction="to"
						onInputChange={this.handleInputChange}
						onCurrencyChange={this.handleCurrencyChange}
					/>
					<div className="btn-wrapper">
						<button type="button" className={`btn ${disabledClass}`}>Exchange</button>
					</div>
				</div>
			</div>
		);
	}

}
