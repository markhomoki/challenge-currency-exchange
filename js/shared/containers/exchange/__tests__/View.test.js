import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from 'app/store';
import { fetchRates } from 'app/actions';
import { GBP, USD } from 'app/types';
import View from '../View';

describe('Exchange.View', () => {

	let wrapper;
	let instance;

	beforeEach(() => {
		wrapper = shallow(<View store={store} />).dive();
		instance = wrapper.instance();
	});

	it('renders correctly', () => {
		const tree = renderer.create(<Provider store={store}><View /></Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('fetches rates', async () => {
		const prevRate = store.getState().rate.value;
		await store.dispatch(fetchRates(USD, GBP));
		const newRate = store.getState().rate.value;
		expect(newRate).not.toBe(prevRate);
	});

	it('swaps the currency pair on clicking swap', () => {
		const { from, to } = instance.state;
		instance.swap();
		expect(instance.state.to).toEqual(from);
		expect(instance.state.from).toEqual(to);
	});

	it('highlights the balance if balance is less than the requested amount', () => {
		instance.handleInputChange(10000, 'from');
		expect(wrapper.find({ testID: 'ExchangeBoxFrom' }).shallow().find('.exchange-item')
			.hasClass('insufficient')).toBe(true);
	});

	it('blocks the button if balance is less than the requested amount', () => {
		instance.handleInputChange(10000, 'from');
		expect(wrapper.find({ testID: 'ExchangeButton' }).prop('disabled')).toBe(true);
	});

	it('blocks the button if amount is 0', () => {
		instance.handleInputChange(0, 'from');

		expect(wrapper.find({ testID: 'ExchangeButton' }).prop('disabled')).toBe(true);
	});

});
