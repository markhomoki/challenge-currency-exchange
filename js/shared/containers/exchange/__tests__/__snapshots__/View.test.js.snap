// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange.View renders correctly 1`] = `
<div
  className="site-inner"
>
  <div
    className="exchange"
  >
    <div
      className="exchange-header"
    >
      <i
        className="ion-close"
      />
      <h1
        className="page-title"
      >
        Exchange
      </h1>
      <i
        className="ion-help-circled"
      />
    </div>
    <div
      className="exchange-item first "
    >
      <div
        className="amount-wrapper"
      >
        <div
          className="currency-wrapper"
        >
          <select
            className="currency-select"
            onChange={[Function]}
            value="USD"
          >
            <option
              value="EUR"
            >
              EUR
            </option>
            <option
              value="GBP"
            >
              GBP
            </option>
            <option
              value="USD"
            >
              USD
            </option>
          </select>
          <i
            className="ion-chevron-down"
          />
        </div>
        <div
          className="digit-input-wrapper"
        >
          <input
            className="digit-input"
            onChange={[Function]}
            placeholder="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        className="balance"
      >
        Balance: 
        $
        143.73
      </span>
    </div>
    <div
      className="separator"
    >
      <button
        className="btn-exchange"
        onClick={[Function]}
        type="button"
      >
        <i
          className="ion-arrow-swap"
        />
      </button>
      <div
        className="rate-wrapper"
      >
        <i
          className="ion-arrow-graph-up-right"
        />
        <span
          className="rate"
        >
          $
          1 = 
          £
          0
        </span>
      </div>
    </div>
    <div
      className="exchange-item alt"
    >
      <div
        className="amount-wrapper"
      >
        <div
          className="currency-wrapper"
        >
          <select
            className="currency-select"
            onChange={[Function]}
            value="GBP"
          >
            <option
              value="EUR"
            >
              EUR
            </option>
            <option
              value="GBP"
            >
              GBP
            </option>
            <option
              value="USD"
            >
              USD
            </option>
          </select>
          <i
            className="ion-chevron-down"
          />
        </div>
        <div
          className="digit-input-wrapper"
        >
          <input
            className="digit-input"
            onChange={[Function]}
            placeholder="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <span
        className="balance"
      >
        Balance: 
        £
        79.14
      </span>
    </div>
    <div
      className="btn-wrapper"
    >
      <button
        className="btn disabled"
        type="button"
      >
        Exchange
      </button>
    </div>
  </div>
</div>
`;
